RISDEV-9134: Draft references ADR (#1273)
digitalservicebund/ris-adm-vwv: doc/adr/0011-cross-referencing-neuris-documents.md

# 11. Cross-referencing neuris documents

Date: 2025-09-12

## Status

Draft

## Context

🚧 The issue motivating this decision, and any context that influences or constrains the decision.🚧 

NeuRIS must support cross-references between documents of heterogeneous types (e.g., caselaw, administrative directives) that are owned by different teams and reside in separate schemas for domain isolation.

Two reference directions must be represented:

- **Active References (Aktivzitierung)**: Explicit links created by a source document to a target.

> - **Passive References (Passivzitierung)**: Implicit links indicating a document has been referenced by another source.

Reviewer: ```suggestion
- **Passive References (Passivzitierung)**: Implicit links indicating that a document is the target of another source's reference.
```

#### Requirements & Constraints:

- Referential integrity must be maintained, even across schema boundaries.
- Streams should have read/write access to references but read-only access to other streams document tables.
- Streams do not reference foreign documents in their stream schema directly.
- The application should be testable in isolation, without relying on a remote shared database.
- CI/CD pipelines must be able to spin up a local environment (e.g., via Docker) with all necessary schemas and tables.

## Decision

🚧 The change that we're proposing or have agreed to implement.🚧 

- To represent references in both directions, even across domains, a central reference table will be introduced: `ris_references` in a shared schema with read and write access for each stream

> - In this shared schema, a central registry table will be introduced which will hold the identifiers for all documents across streams.

Author: I thought that these meta-data (e.g. title) would rather be in the central registry table from the shared schema where all documents are indexed. The `ris_references` table would hold only the relationships.

> - In this shared schema, a central registry table will be introduced which will hold the identifiers for all documents across streams.

Reviewer: The idea really was to attach it to the reference records.

> - Each stream-specific schema will maintain its own documents table for storing its document data, and will have read-only access to other streams documents tables.

Reviewer: The layer of "ref view" is missing

Author: Yes, I was not sure if we would have views if we cannot maintain the ref integrity via fk?

Reviewer: Right, we don't know, yet.
Therefore: let's have this as the current decision - and adjust once we learn more. :+1:
